speaker_verification/denoise.py

```diff
@@ -1,130 +1,44 @@
-            a = 4
-        if SNR > 20:
-            a = 1
-    return a
-
-
-def find_index(x_list):
-    index_list = []
-    for i in range(len(x_list)):
-        if x_list[i] < 0:
-            index_list.append(i)
-    return index_list
-
-
-def denoise(audio_file, input_path, output_path):
-    file_name = os.path.basename(audio_file).split(".")[0]
-    input_file_path = os.path.join(input_path, audio_file)
-    output_file_path = os.path.join(output_path, file_name)
-
-    with wave.open(input_file_path, 'r') as f:
-        # (nchannels, sampwidth, framerate, nframes, comptype, compname)
-        params = f.getparams()
-        nchannels, sampwidth, framerate, nframes = params[:4]
-        str_data = f.readframes(nframes)
+import noisereduce as nr
+import numpy as np
+from scipy.io import wavfile
 
-    x = np.frombuffer(str_data, dtype=np.short)
-    len_ = 20 * framerate // 1000
-    len1 = len_ * 50 // 100
-    # The percentage of the frame that the window overlaps
-    len2 = len_ - len1
-    Thres = 3
-    Expnt = 2.0
-    beta = 0.002
-    G = 0.9
-    win = np.hamming(len_)
-    # normalization gain for overlap+add with 50% overlap
-    winGain = len2 / sum(win)
 
-    # Noise magnitude calculations - assuming that the first 5 frames is noise/silence
-    nFFT = 2 * 2 ** (nextpow2(len_))
-    noise_mean = np.zeros(nFFT)
+def denoise(input_file: str) -> (np.ndarray, int):
+    """
+    Denoise audio file using noisereduce.
+    """
 
-    j = 0
-    for k in range(1, 6):
-        noise_mean = noise_mean + abs(np.fft.fft(win * x[j: j + len_], nFFT))
-        j = j + len_
-    noise_mu = noise_mean / 5
+    # TODO: Read other audio types beside .wav
+    # Read audio data
+    rate, data = wavfile.read(input_file)
+    data = data.astype(np.float64)
 
-    # --- allocate memory and initialize various variables
-    k = 1
-    img = 1j
-    x_old = np.zeros(len1)
-    Nframes = len(x) // len2 - 1
-    xfinal = np.zeros(Nframes * len2)
+    # Section of data that is noisy.
+    # Since we don't know, just select the whole data
+    noise_clip = data
 
-    # Start Processing
-    for n in range(0, Nframes):
-        # Windowing
-        insign = win * x[k - 1: k + len_ - 1]
-        # compute fourier transform of a frame
-        spec = np.fft.fft(insign, nFFT)
-        # compute the magnitude
-        sig = abs(spec)
+    # Perform noise reduction
+    noise_reduced_data = nr.reduce_noise(audio_clip=data, noise_clip=noise_clip, verbose=False)
 
-        # save the noisy phase information
-        theta = np.angle(spec)
-        SNRseg = 10 * np.log10(
-            np.linalg.norm(sig, 2) ** 2 / np.linalg.norm(noise_mu, 2) ** 2
-        )
+    return rate, noise_reduced_data
 
-        if Expnt == 1.0:
-            alpha = berouti1(SNRseg)
-        else:
-            alpha = berouti(SNRseg)
 
-        sub_speech = sig ** Expnt - alpha * noise_mu ** Expnt
+def store_wav_audio(wav_data: np.ndarray, rate: int, output_file: str):

```

Comment:
We don't want to be writing the denoised data to file as we can't actually store and audio data to the database. Instead we should be altering the current ingest of audio data, add a denoise step, and then save the data sample to our database. I'd look at adding the denoise function around [here in the code](https://github.com/OnTrack-UG-Squad/speaker-verification/blob/development/speaker_verification/deep_speaker/audio.py#L48), you could refactor your `denoise` function to accept the `audio` array from `audio_read` and then return a denoised array.

Reply:
@MoeHdaib I think you resolved this issue right?